### PR TITLE
Rust sample app

### PR DESCRIPTION
**Pull request by catenocrypt**
## Description

Sample application in Rust language.

## How to test

- build and run sample app, according to the README.md.
- Check that sample is executed in the 'sample app' CI

image

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist

- [ ] Implement proper destructor-cleanup of interface data types (TWData, ...) 
- [x] Create pull request as draft initially, unless its complete.
- [x] Add tests to cover changes as needed.
- [x] Update documentation as needed.
- [x] If there is a related Issue, mention it in the description.

If you're adding a new blockchain

- [ ] I have read the [guidelines](https://developer.trustwallet.com/wallet-core/newblockchain#integration-criteria) for adding a new blockchain.

### CODE REVIEW

File: samples/rust/src/build.rs [hunk 1]
```diff
@@ -0,0 +1,17 @@
+static WALLET_CORE_PROJECT_DIR: &str = "../../build";
+    // libs to link with
+    static LIBS: [&str; 3] = ["TrustWalletCore", "TrezorCrypto", "protobuf"];
```
Comment (hewigovens) [inline samples/rust/src/build.rs:3]:
```suggestion
// libs to link with
static LIBS: [&str; 3] = ["TrustWalletCore", "TrezorCrypto", "protobuf"];
```

File: samples/rust/src/main.rs [hunk 2]
```diff
@@ -0,0 +1,27 @@
+pub mod walletcore;
```
Comment (hewigovens) [inline samples/rust/src/main.rs:1]: can you add license header?

File: samples/rust/src/walletcore.rs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    wrapped: *const i8
+    // TODO delete when destructing with TWStringDelete
+}
+
+pub trait FromString {
+    fn from_str(s: &str) -> Self;
+}
+
+impl FromString for TWString {
+    //fn from_str(s: &str) -> Result<Self, Self::Err> {
```
Comment (Milerius) [inline samples/rust/src/walletcore.rs:27]: Let's remove this comment for now

File: samples/rust/src/walletcore.rs [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+        let ptr = unsafe { TWStringCreateWithUTF8Bytes(cstring.as_ptr()) };
+        return TWString { wrapped: ptr };
+    }
+}
+
+impl ToString for TWString {
+    fn to_string(&self) -> String {
+        let s1 = unsafe { TWStringUTF8Bytes(self.wrapped) };
+        let c_str: &CStr = unsafe { CStr::from_ptr(s1) };
+        let str_slice: &str = c_str.to_str().unwrap();
```
Comment (Milerius) [inline samples/rust/src/walletcore.rs:39]: Better to use a `Result` and `?` operator or rename the function `to_string_unsafe`

Note in rust you can implement the traits `From` `TryFrom` for such operation: https://doc.rust-lang.org/stable/std/convert/trait.TryFrom.html

Comment (catenocrypt) [reply samples/rust/src/walletcore.rs:71]: I skip this in this sample

File: .github/workflows/linux-sampleapp-ci.yml [hunk 5]
```diff
@@ -41,6 +41,7 @@ jobs:
         CXX: /usr/bin/clang++
     - name: Build
       run: |
+        echo $nproc
```
Comment (catenocrypt) [inline .github/workflows/linux-sampleapp-ci.yml:44]: Testing only, does not work, remove

File: samples/rust/src/wc_proto/README.md [hunk 6]
```diff
@@ -0,0 +1,2 @@
+Source files generated from the protobuf .proto files will come here, during build time.
```
Comment (hewigovens) [inline samples/rust/src/wc_proto/README.md:1]: I think you can create this folder during cargo build? no need to add a placeholder

File: samples/rust/src/walletcore_iface.rs [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+    fn TWHDWalletGetAddressForCoin(wallet: *const u8, coin: u32) -> *const u8;
+    fn TWHDWalletGetKeyForCoin(wallet: *const u8, coin: u32) -> *const u8;
+
+    fn TWAnySignerSign(input: *const u8, coin: u32) -> *const u8;
+
+    fn TWMnemonicIsValid(mnemonic: *const u8) -> bool;
+}
+
+// type utilities
+pub struct TWString {
```
Comment (hewigovens) [inline samples/rust/src/walletcore_iface.rs:43]: Move `TWString` and `TWData` code to separate files, we can reuse them later (the reset can be generated)

File: samples/rust/src/wc_proto/README.md (hunk 6 above)
Comment (catenocrypt) [reply samples/rust/src/wc_proto/README.md]: +1

File: samples/rust/src/walletcore_iface.rs (hunk 7 above)
Comment (catenocrypt) [reply samples/rust/src/walletcore_iface.rs:42]: Good insight; I separated it into interface code that could be generated and some extra traits, for TWString and TWData.

**Review by hewigovens — APPROVED** [review]

File: samples/rust/src/walletcore_iface.rs (hunk 7 above)
Comment (Aminkavoos) [reply samples/rust/src/walletcore_iface.rs:18]: ****

**Review by Milerius — APPROVED** [review]

**Review by Milerius — APPROVED** [review]

**Pull request merged**